src/liballoc/string.rs

```diff
@@ -1031,6 +1031,57 @@ impl String {
         ch
     }
 
+    /// Retains only the characters specified by the predicate.
+    ///
+    /// In other words, remove all characters `c` such that `f(c)` returns `false`.
+    /// This method operates in place and preserves the order of the retained
+    /// characters.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(string_retain)]
+    ///
+    /// let mut s = String::from("f_o_ob_ar");
+    ///
+    /// s.retain(|c| c != '_');
+    ///
+    /// assert_eq!(s, "foobar");
+    /// ```
+    #[inline]
+    #[unstable(feature = "string_retain", issue = "0")]
+    pub fn retain<F>(&mut self, mut f: F)
+        where F: FnMut(char) -> bool
```

Comment:
In [HashMap::retain](https://github.com/rust-lang/rust/blob/master/src/libstd/collections/hash/map.rs#L1243), we used `F: FnMut(&K, &mut V) -> bool`, in case we need to mutate the value in iteration. @alexcrichton  also thinks it is better if we can improve  [Vec::retain](https://github.com/rust-lang/rust/blob/master/src/liballoc/vec.rs#L805) to be `F: FnMut(&mut T) -> bool`, though it is too late to do this. So I guess it could be better if we use `F: FnMut(&mut char)->bool`.